006_plug_and_play_prior/toto.ipynb

```diff
@@ -0,0 +1,705 @@
+    "y_data = Y_generation(X_generation = im).sample()\n",
+    "\n",
+    "fig = plt.figure(figsize = (10, 10))\n",
+    "ax1 = fig.add_subplot(1, 2, 1)\n",
+    "ax1.imshow(im, cmap = \"gray\")\n",
+    "ax1.axis(\"off\")\n",
+    "ax1.set_title(\"Original image\")\n",
+    "ax2 = fig.add_subplot(1, 2, 2)\n",
+    "ax2.imshow(np.reshape(y_data, (N, N)), cmap = \"gray\")\n",
+    "ax2.set_title(\"Observation\")\n",
+    "ax2.axis(\"off\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Loading the Plug-and-Play regularizer\n",
+    "\n",
+    "We first load the pre-trained denoiser which implicitly defines a regularizer. It has been trained at noise level $s=2.25$."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "prior_denoiser = \"fine\" # \"sn_dncnn\" or \"fine\"\n",
+    "device = \"cpu\" # \"cpu\" or \"cuda:0\"\n",
+    "if prior_denoiser == \"sn_dncnn\":\n",
+    "    sys.path.append(\"./pnp_denoiser/sn_dncnn/\")\n",
+    "    from ryu_utils.utils import load_model\n",
+    "    from denoiser import pytorch_denoiser_residual\n",
+    "    if device == \"cpu\":\n",
+    "        cuda = False\n",
+    "    else:\n",
+    "        cuda = True\n",
+    "    s = 5 # s = 5, 15, 40\n",
+    "    model = load_model(model_type = \"RealSN_DnCNN\", sigma = s, \\\n",
+    "                       device = device, cuda = cuda, \\\n",
+    "                       path = \"pnp_denoiser/sn_dncnn/Pretrained_models/\")\n",
+    "    Ds = lambda x, z : pytorch_denoiser_residual(x, model, device)\n",
+    "\n",
+    "elif prior_denoiser == \"fine\":\n",
+    "    # Firmly non-expansive network. \n",
+    "    sys.path.append(\"./pnp_denoiser/sn_dncnn/\")\n",
+    "    sys.path.append(\"./pnp_denoiser/fine/\")\n",
+    "    from utils_terris_dncnn import load_dncnn_weights_2\n",
+    "    from denoiser import pytorch_denoiser\n",
+    "    if device == \"cpu\":\n",
+    "        cuda = False\n",
+    "    else:\n",
+    "        cuda = True\n",
+    "    s = 2.25 # s = 2.0, 2.25, 2.5\n",
+    "    n_ch = 1\n",
+    "    if s == 2.5:\n",
+    "        ljr = 0.005\n",
+    "    elif s == 2.25 or s == 2.0:\n",
+    "        ljr = 0.002\n",
+    "    model = load_dncnn_weights_2(n_ch, s, ljr, device, path = 'pnp_denoiser/fine/ckpts/finetuned/')\n",
+    "    Ds = lambda x, delta : pytorch_denoiser(x, model, device)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Bayesian problem formulation\n",
+    "\n",
+    "Now we are ready to define the Bayesian problem in CUQIpy."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Sample 250 / 250\n"
+     ]
+    }
+   ],
+   "source": [
+    "X = RegularizedGaussian(np.zeros(N*N), gamma2, proximal = Ds)\n",
```

Comment:
Please explain that the denoiser is to be interpreted as a proximal mapping and RegularizedGaussian can be defined with arbitrary proximal mapping functions, so we pass it as the proximal input argument.